src/number_theory/pell.lean

```diff
@@ -241,6 +144,220 @@ lemma x_neg (a : solution₁ d) : (-a).x = -a.x := rfl
 @[simp]
 lemma y_neg (a : solution₁ d) : (-a).y = -a.y := rfl
 
+/-- A solution has `x ≠ 0`. -/
+lemma x_ne_zero (h₀ : 0 < d) (a : solution₁ d) : a.x ≠ 0 :=
+begin
+  intro hx,
+  have h : 0 ≤ d * a.y ^ 2 := mul_nonneg h₀.le (sq_nonneg _),
+  rw [a.prop_y, hx, sq, zero_mul, zero_sub] at h,
+  exact not_le.mpr (neg_one_lt_zero : (-1 : ℤ) < 0) h,
+end
+
+/-- A solution with `x > 1` must have `y ≠ 0`. -/
+lemma y_ne_zero_of_one_lt_x {a : solution₁ d} (ha : 1 < a.x) : a.y ≠ 0 :=
+begin
+  intro hy,
+  have prop := a.prop,
+  rw [hy, sq (0 : ℤ), zero_mul, mul_zero, sub_zero] at prop,
+  exact lt_irrefl _ (((one_lt_sq_iff $ zero_le_one.trans ha.le).mpr ha).trans_eq prop),
+end
+
+/-- A solution with `x = 1` is trivial. -/
+lemma eq_one_of_x_eq_one (h₀ : 0 < d) {a : solution₁ d} (ha : a.x = 1) : a = 1 :=
+begin
+  have hy : a.y = 0,
+  { have prop := a.prop.symm,
+    rw [ha, ← sub_eq_zero] at prop,
+    ring_nf at prop,
+    nlinarith, },
+  ext; simp [ha, hy],
+end
+
+/-- A solution is `1` or `-1` if and only if `y = 0`. -/
+lemma eq_one_or_neg_one_iff_y_eq_zero {a : solution₁ d} : a = 1 ∨ a = -1 ↔ a.y = 0 :=
+begin
+  refine ⟨λ H, H.elim (λ h, by simp [h]) (λ h, by simp [h]), λ H, _⟩,
+  have prop := a.prop,
+  rw [H, sq (0 : ℤ), mul_zero, mul_zero, sub_zero, sq_eq_one_iff] at prop,
+  refine prop.elim (λ h, or.inl $ by ext; simp [h, H]) (λ h, or.inr $ by ext; simp [h, H]),
+end
+
+/-- The set of solutions with `x > 0` is closed under multiplication. -/
+lemma x_mul_pos_of_x_pos (h₀ : 0 < d) {a b : solution₁ d} (ha : 0 < a.x) (hb : 0 < b.x) :
+  0 < (a * b).x :=
+begin
+  simp only [x_mul],
+  refine neg_lt_iff_pos_add'.mp (abs_lt.mp _).1,
+  rw [← abs_of_pos ha, ← abs_of_pos hb, ← abs_mul, ← sq_lt_sq, mul_pow a.x, a.prop_x, b.prop_x,
+      ← sub_pos],
+  ring_nf,
+  positivity,
+end
+
+/-- The set of solutions with `x` and `y` positive is closed under multiplication. -/
+lemma x_mul_pos_and_y_mul_pos_of_x_pos_of_y_pos (h₀ : 0 < d) {a b : solution₁ d} (hax : 0 < a.x)
+  (hay : 0 < a.y) (hbx : 0 < b.x) (hby : 0 < b.y) :
+  0 < (a * b).x ∧ 0 < (a * b).y :=
+begin
+  simp only [x_mul, y_mul],
+  split; positivity,
+end
+
+/-- If `(x, y)` is a solution with `x` and `y` positive, then all powers of it have positive `x`,
+and the sign of `y` agrees with that of the exponent. -/
+lemma x_pow_pos_of_x_pos_of_y_pos (h₀ : 0 < d) {a : solution₁ d} (hax : 0 < a.x) (hay : 0 < a.y)
+  (n : ℤ) :
+  0 < (a ^ n).x ∧
+  (if 0 < n then 0 < (a ^ n).y else if n < 0 then (a ^ n).y < 0 else (a ^ n).y = 0) :=
```

Comment:
Would it be easier to state this in terms of `int.sign`?

Reply:
That sounds good to me. It's also pretty unusual to state a lemma with an `and` as a conclusion; we'd almost always split it into two lemma instead.

Reply:
I don't think a disjunction makes sense here, considering that the two sides aren't even of the same form. It would be one thing if you phrased everything in terms of `0 < a.x ∧ 0 < a.y`, but here the two conclusions really should be separate lemmas I feel.

Reply:
OK.